src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Android.cs

```diff
@@ -0,0 +1,390 @@
+        * The older locations still exist (at least the `/system/usr/share/zoneinfo` one) but they won't be updated.
+        */
+        private sealed class AndroidTzData
+        {
+            [StructLayout(LayoutKind.Sequential, Pack=1)]
+            private unsafe struct AndroidTzDataHeader
+            {
+                public fixed byte signature[12]; // "tzdata2012f\0"
+                public int indexOffset;
+                public int dataOffset;
+                public int finalOffset;
+            }
+
+            [StructLayout(LayoutKind.Sequential, Pack=1)]
+            private unsafe struct AndroidTzDataEntry
+            {
+                public fixed byte id[40];
+                public int byteOffset;
+                public int length;
+                public int unused; // Was raw GMT offset; always 0 since tzdata2014f (L).
+            }
+
+            private string[]? _ids;
+            private int[]? _byteOffsets;
+            private int[]? _lengths;
+
+            public AndroidTzData()
+            {
+                ReadHeader(GetTimeZoneDirectory() + _timeZoneFileName);
+            }
+
+            private unsafe void ReadHeader(string tzFilePath)
+            {
+                int size = Math.Max(Marshal.SizeOf(typeof(AndroidTzDataHeader)), Marshal.SizeOf(typeof(AndroidTzDataEntry)));
+                var buffer = new byte[size];
+                AndroidTzDataHeader header = ReadAt<AndroidTzDataHeader>(tzFilePath, 0, buffer);
+
+                header.indexOffset = NetworkToHostOrder(header.indexOffset);
+                header.dataOffset = NetworkToHostOrder(header.dataOffset);
+
+                // tzdata files are expected to start with the form of "tzdata2012f\0" depending on the year of the tzdata used which is 2012 in this example
+                sbyte* s = (sbyte*)header.signature;
+                string magic = new string(s, 0, 6, Encoding.ASCII);
+
+                if (magic != "tzdata" || header.signature[11] != 0)
+                {
+                    var b = new StringBuilder();
+                    for (int i = 0; i < 12; ++i) {
+                        b.Append(' ').Append(((byte)s[i]).ToString("x2"));
+                    }
+
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_BadTZHeader, tzFilePath, b.ToString()));
+                }
+
+                ReadIndex(tzFilePath, header.indexOffset, header.dataOffset, buffer);
+            }
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2087:UnrecognizedReflectionPattern",
+                Justification = "Implementation detail of Android TimeZone")]
+            private unsafe T ReadAt<T>(string tzFilePath, long position, byte[] buffer)
+                where T : struct
+            {
+                int size = Marshal.SizeOf(typeof(T));
+                if (buffer.Length < size)
+                {
+                    throw new InvalidOperationException(SR.InvalidOperation_BadBuffer);
```

Comment:
Do we really need a resource exception for this case? Could we just live with a Debug.Assert? since it would be our bug if the buffer wasn't big enough.

Reply:
Modified to be `Debug.Assert(buffer.Length >= size);`
In general, when would we want resource exceptions vs Debug.Assert?

Reply:
In general, if it is a check that ensures our internal code is doing the right thing - and the check can only fail if we have a bug ourselves - use Debug.Assert. For all other cases, e.g. where a public caller did something wrong, or a file had bad data, etc, use an exception.

Reply:
@eerhardt `Debug.Assert` suggestion is fine provided that the test suite runs in Debug builds, otherwise it's an effective no-op and can hide bugs. If there's no guarantee that tests will be built and executed in Debug configuration, such checks should remain unconditional runtime ones.

Reply:
Yes, we run all tests in both Debug and Release builds.

Reply:
Good, that helps. However, I think it's not a good idea to use `Debug.Assert` (or `assert(3)` in native code) because it provides a false sense of safety - even if the tests are executed in Debug mode, the invariant **may** be broken in execution of actual application's production build. If the assertion is not checked in this case, it may still lead to serious issues, internal bug or not. This is especially important regarding buffer size checks, variable initialization checks, method/function argument checks. Of course, not everything should be checked at every step of the way because we'd spend as much time checking as executing the actual code, but things like pointers or buffers should always be validated.

Reply:
I agree that buffer sizes that are outside of the immediate code control should be always validated. It is not the case here. This is private method, and the buffer size is controller by the immediate caller. You can find many similar cases in CoreLib where the buffer sizes in situations like this one are checked by asserts only.

Also, I expect that the code will change significantly after the feedback gets incorporated, and this point will be moot.